prometheus_hardware_exporter/collectors/redfish.py

```diff
@@ -1,24 +1,41 @@
 """Redfish collector."""
 from logging import getLogger
-from typing import Any, Dict, List, Optional
+from typing import Any, Callable, Dict, List, Optional
 
 import redfish
 import redfish_utilities
-from redfish.rest.v1 import HttpClient, InvalidCredentialsError, SessionCreationError
+from cachetools.func import ttl_cache
+from redfish.rest.v1 import (
+    HttpClient,
+    InvalidCredentialsError,
+    RetriesExhaustedError,
+    SessionCreationError,
+)
 
 logger = getLogger(__name__)
 
 
+# pylint: disable=R0903
+
+
 class RedfishHelper:
     """Helper function for redfish."""
+    def __init__(self, discover_cache_ttl: int) -> None:
+        """Initialize disover method with TTL value."""
+        self.discover = get_discover(discover_cache_ttl)
 
-    def get_sensor_data(self, host: str, username: str, password: str) -> Dict[str, List]:
+    def get_sensor_data(self, login_info: dict, timeout: int, max_retry: int) -> Dict[str, List]:
```

Comment:
not-blocking: I think it would be more readable to leave `host`, `username` and `password ` as arguments, instead of hiding it to one single argument.

Reply:
The reason behind this was because there was a pylint error with too many arguments for the method (6 including `self`, it should be below 5). So I thought I'd collect all the redfish login related vars into one dict and pass that to the method.

Should I disable the pylint error instead and pass the vars separately?

Reply:
I am agree with Robert, `login_info` doesn't make things clear. We can try to ignore pylint here.

Reply:
Since both of you are suggesting this, I'll revert my change, separate all the args again and add a "pylint disable" comment. IMO I think it's best to keep the number of arguments low to avoid too much complexity in the function and to bundle related arguments in a data structure so that it's easier to understand, maintain and test overall.

@rgildein - IMO it's good to keep the default number of max arguments (6) and not ignore this error by default from the template. Ideally 3-4 args would be good for a function and if it needs more, it's probably a sign that the function is doing too much work and we can split into another function. For this specific case, I can add the comment to disable to disable the pylint error.

Reply:
> IMO I think it's best to keep the number of arguments low to avoid too much complexity in the function and to bundle related arguments in a data structure so that it's easier to understand, maintain and test overall.

If you wrap 3 arguments into single one and than unwrap it and use all those arguments + other in single call, it's not cleaner, easier to tests and deferentially not providing lower complexity.

```python
def foo(a: Dict[str, Any], b: int, c: int) -> None:
    a1 = a["a1"]
    a2 = a["a2"]
    a3 = a["a3"]
    foo2(a1, a2, a3, b, c)
```

I'm not sure that 6 is always enough, since we are using a lot 3-th party staff and our functions sometimes needs too many arguments, because they are forwarding them to the external function. I think that 9 will make much more sense. I'm doing lot of reviews and everyday I'm saying usage of pylint disable for this.

Btw, never use force push after review, because now I could not check what did you changed from my last review. Thanks

Reply:
I understand the situation in your example where encapsulating all the args into a single one might not always have lesser complexity when it's easier to pass them all along separately. 

But I think as a general practice if the function is taking that many number of arguments (>6), it would be a bit complex to make sure to pass all the right args to the function each time you call it. Even during testing, we need to send all these args. But at the end of the day, it all depends on the exact situation :)

> Btw, never use force push after review, because now I could not check what did you changed from my last review.

Ah, it's a force of habit my bad. I don't want to mess up the log with lots of useless comments but I also understand we can just squash them before merging. Will make sure to not force push from now :+1:

Reply:
> I understand the situation in your example where encapsulating all the args into a single one might not always have lesser complexity when it's easier to pass them all along separately.

I will not have any issue if you do 
```python
def foo(foo_arguments) -> None:
    """...
    params:
        foo_arguments: arguments for foo2
    """
    foo2(**foo_arguments)
```
Or simply
```python
def foo(*args, **kwargs) -> None:
    ...
    foo2(*args, **kwargs)
```

But in your case, when you never used `login_info` anywhere and simply pass those along with `timeout` and `max_retry` to sub-function, same function. Like this you only creates potential issue that someone will define `login_info` wrongly and got simple KeyError instead of `TypeError: 'a' is an invalid keyword argument for redfish.redfish_client`.

When you are using good IDE, it will tells you that you are missing e.g. `host` or that you provide wrong type for `host`, but when you do `login_info` it will tells you nothing.

Reply:
After the latest changes, all the values are being sent to `RedfishHelper` as an instance of the `Config` object containing the login args (host, username, password) and the ttl value for the cache. This eliminates the need to send it all separately and also makes sure there is a tight coupling between the related data elements. It's also easier to instantiate the class now and prevents declaring an arbitrary `login_info` which isn't being used anywhere else.